node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/ArchiveUris.java

```diff
@@ -0,0 +1,86 @@
+import java.util.regex.Pattern;
+
+/**
+ * Thread safe class to get and set archive URI for given tenants. Archive URIs are stored in ZooKeeper so that
+ * nodes within the same tenant have the same archive URI from all the config servers.
+ *
+ * @author freva
+ */
+public class ArchiveUris {
+
+    private static final Logger log = Logger.getLogger(ArchiveUris.class.getName());
+    private static final Pattern validUriPattern = Pattern.compile("[a-z0-9]+://(?:(?:[a-z0-9]+(?:-[a-z0-9])*)+/)+");
+    private static final Duration cacheTtl = Duration.ofMinutes(1);
+
+    private final CuratorDatabaseClient db;
+    private final CachedSupplier<Map<TenantName, String>> archiveUris;
+
+    public ArchiveUris(CuratorDatabaseClient db) {
+        this.db = db;
+        this.archiveUris = new CachedSupplier<>(db::readArchiveUris, cacheTtl);
+    }
+
+    /** Returns the current archive URI for each tenant */
+    public Map<TenantName, String> getArchiveUris() {
+        return archiveUris.get();
+    }
+
+    /** Returns the archive URI to use for given tenant */
+    public Optional<String> archiveUriFor(TenantName tenant) {
+        return Optional.ofNullable(archiveUris.get().get(tenant));
+    }
+
+    /** Returns the archive URI to use for given node */
+    public Optional<String> archiveUriFor(Node node) {
+        return node.allocation().map(Allocation::owner)
+                .flatMap(app -> archiveUriFor(app.tenant())
+                        .map(uri -> {
+                            StringBuilder sb = new StringBuilder(100).append(uri)
+                                    .append(app.application().value()).append('/')
+                                    .append(app.instance().value()).append('/');
+
+                            for (char c: node.hostname().toCharArray()) {
+                                if (c == '.') break;
+                                sb.append(c);
+                            }
+
+                            return sb.append('/').toString();
+                        }));
+    }
+
+    /** Set the docker image for nodes of given type */
+    public void setArchiveUri(TenantName tenant, Optional<String> archiveUri) {
+        try (Lock lock = db.lockArchiveUris()) {
+            Map<TenantName, String> archiveUris = new TreeMap<>(db.readArchiveUris());
+            if (Optional.ofNullable(archiveUris.get(tenant)).equals(archiveUri)) return; // No change
+
+            archiveUri.map(ArchiveUris::normalizeUri).ifPresentOrElse(uri -> archiveUris.put(tenant, uri),
+                                                                      () -> archiveUris.remove(tenant));
+            db.writeArchiveUris(archiveUris);
+            this.archiveUris.refresh(); // Throw away current cache
+            log.info("Set archive URI for " + tenant + " to " + archiveUri.orElse(null));
+        }
+    }
+
+    static String normalizeUri(String uri) {
+        if (!uri.endsWith("/")) uri = uri + "/";
```

Comment:
Any reason to not use `java.net.URI` for the URIs? Then you get (some) validation and normalization for free.

Reply:
`java.net.URI` is just way to permissive, f.ex. all of the invalid URIs from my test are legal `java.net.URI`. With the regex above we can limit it down just to the format we want it.

Reply:
Sure, but I still think using `java.net.URI` for something that is an URI is better.

Reply:
That is true, decided against it since it is only used in REST API, and even before that, the URI needs to be extended, so we would have to do a lot of `URI::create` and `URI::toString` to do anything with it.